api/v1beta1/module_webhook.go

```diff
@@ -0,0 +1,108 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"fmt"
+	"k8s.io/apimachinery/pkg/runtime"
+	"regexp"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var modulelog = logf.Log.WithName("module-resource")
+
+func (r *Module) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/validate-kmm-sigs-x-k8s-io-v1beta1-module,mutating=false,failurePolicy=fail,sideEffects=None,groups=kmm.sigs.x-k8s.io,resources=modules,verbs=create;update,versions=v1beta1,name=vmodule.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &Module{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *Module) ValidateCreate() error {
+	modulelog.Info("validate create", "name", r.Name)
+
+	if err := r.validateKernelMappingRegex(); err != nil {
+		return err
+	}
+
+	return r.validateModuleName()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *Module) ValidateUpdate(_ runtime.Object) error {
+	modulelog.Info("validate update", "name", r.Name)
+
+	if err := r.validateKernelMappingRegex(); err != nil {
+		return err
+	}
+
+	return r.validateModuleName()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *Module) ValidateDelete() error {
+	return nil
+}
+
+func (r *Module) validateKernelMappingRegex() error {
+	if r.Spec.ModuleLoader.Container.KernelMappings == nil {
+		return nil
+	}
+
+	container := r.Spec.ModuleLoader.Container
+
+	for idx, km := range container.KernelMappings {
+		if _, err := regexp.Compile(km.Regexp); err != nil {
+			return fmt.Errorf("invalid regexp at index %d: %v", idx, err)
+		}
+
+		if len(container.ContainerImage) == 0 && len(km.ContainerImage) == 0 {
+			return fmt.Errorf("missing spec.moduleLoader.container.kernelMappings[].containerImage")
+		}
+	}
+
+	return nil
+}
+
+func (r *Module) validateModuleName() error {
+	modprobe := r.Spec.ModuleLoader.Container.Modprobe
+
+	if len(modprobe.ModuleName) > 0 {
+		return nil
+	}
+
+	if modprobe.RawArgs == nil {
+		return fmt.Errorf("missing spec.moduleLoader.container.modprobe.moduleName")
+	}
+
+	if len(modprobe.RawArgs.Load) == 0 {
+		return fmt.Errorf("missing spec.moduleLoader.container.modprobe.rawArgs.load")
+	}
+
+	if len(modprobe.RawArgs.Unload) == 0 {
+		return fmt.Errorf("missing spec.moduleLoader.container.modprobe.rawArgs.unload")
+	}
```

Comment:
```suggestion
	if len(modprobe.RawArgs.Load) == 0 || len(modprobe.RawArgs.Unload) == 0 {
		return fmt.Errorf("when using spec.moduleLoader.container.modprobe.rawArgs, both load and unload fields should be specified")
	}
```

I think that if you are using rawArgs you must specify both:
1. How to load
2. How to unload

@qbarrand Please keep me honest here.

Reply:
I prefer sub-sequent `if` statements instead of one with multiple conditions... The validation won't pass if there is one missing so it's checking the same, isn't it?

Reply:
You are right, the validation will fail anyway, the error message is a bit less explicit though. I think it makes sense to group them together as they are treated as a pair (both are set or both unset) but if you insist on keeping 2 separate `if`s I won't object.